### DISCUSSION TITLE
Inference for ensemble

### DISCUSSION THREAD

**Discussion by LSnyd** [MIC-DKFZ/nnUNet#1004] [Q&A]
Hi all, 
I've ran the nn-Unet on my dataset and it returned an ensemble between 3D_fullres and 3D_cascade as the best model. 
When I am running the inference, I am only getting the predictions for the 3D_fullres and 3D_cascade model separately. There is no prediction created for the ensemble of both. Is that correct? 
The ensemble folder just includes an empty "not_postprocessed" folder and a  postprocessing.json file after inference.

**Comment by FabianIsensee** [comment]
You need to use the nnUNet_ensemble command for that.
nnUNet_find_best_configuration actually prints the inference command you should use. Just copy that :-)

**Reply by LSnyd** [reply]
Hi @FabianIsensee,
I copied and used all three lines: inference for model1, inference for model2 and last the ensemble comment as it was displayed after the find_best_config (see below). I added the folder paths. Howeverm I am getting predictions for model1 and model2, but not for the ensemble. I am supposed to be getting output images for the ensemble as well, correct? 

nnUNet_predict -i _folder test images_ -o _output folder model1_ -tr nnUNetTrainerV2 -ctr nnUNetTrainerV2CascadeFullRes -m 3d_fullres -p nnUNetPlansv2.1 -t Task001

nnUNet_predict -i _folder test images_ -o _output folder model2_ -tr nnUNetTrainerV2 -ctr nnUNetTrainerV2CascadeFullRes -m 3d_cascade_fullres -p nnUNetPlansv2.1 -t Task001

nnUNet_ensemble -f _output folder model1_ _output folder model2_ -o _output folder ensemble_ -pp /../nnUNet/nnUNet_trained_models/nnUNet/ensembles/Task001/ensemble_3d_fullres__nnUNetTrainerV2__nnUNetPlansv2.1--3d_cascade_fullres__nnUNetTrainerV2CascadeFullRes__nnUNetPlansv2.1/postprocessing.json

**Comment by LSnyd** [comment]
@FabianIsensee can you confirm that there should be extra segmentation outputs for the inference of the ensemble?

**Comment by FabianIsensee** [comment]
Yes but you also need to look in the correct folder. You just provided `-o output folder ensemble` so the ensemble output is in whatever folder you cwd was during execution of the command.

**Reply by LSnyd** [reply]
Hi @FabianIsensee, I am quite sure that I am checking the correct folder. I just displayed it in a more clear way without the actual folder path. 
The folder for model1 carries the segmented nifti files, the folder for model2 as well, but the ensemble output folder only includes post_processing.json and an empty folder called not_postprocessed. 

When I am running the last command, I am getting the following output: 

```
Please cite the following paper when using nnUNet:

Isensee, F., Jaeger, P.F., Kohl, S.A.A. et al. "nnU-Net: a self-configuring method for deep learning-based biomedical image segmentation." Nat Methods (2020). https://doi.org/10.1038/s41592-020-01008-z


If you have questions or suggestions, feel free to open an issue at https://github.com/MIC-DKFZ/nnUNet

Postprocessing...
```

Then it finished more or less immediately without throwing any errors.

**Reply by FabianIsensee** [reply]
Did you run the two prediction commands with the `-z` flag? There should be .npz and .pkl files in those (these are used by the ensembling)

**Reply by LSnyd** [reply]
No there was no -z flag included. I'll try it and get back to you again. Thank you!

**Reply by FabianIsensee** [reply]
then I need to improve that command ;-) thanks

**Reply by LSnyd** [reply]
Worked with the -z flag, thanks!
